lib/IDE/SwiftSourceDocInfo.cpp

```diff
@@ -189,10 +189,24 @@ bool CursorInfoResolver::walkToExprPre(Expr *E) {
       }
     }
     auto IsProperCursorLocation = E->getStartLoc() == LocToResolve;
-    // Handle cursor placement between try and ! in ForceTryExpr.
-    if (auto *FTE = dyn_cast<ForceTryExpr>(E)) {
-      IsProperCursorLocation = LocToResolve == FTE->getExclaimLoc() ||
-      IsProperCursorLocation;
+    // Handle cursor placement after Try, ForceTry and OptionalTry Expr.
+    if (auto *ATE = dyn_cast<AnyTryExpr>(E)) {
+      auto CheckLocation = [&IsProperCursorLocation, this](SourceLoc Loc) {
+        IsProperCursorLocation = Loc == LocToResolve || IsProperCursorLocation;
+      };
+      auto TryLoc = ATE->getTryLoc();
+      auto TryEndLoc = TryLoc.getAdvancedLocOrInvalid(getKeywordLen(tok::kw_try));
+      CheckLocation(TryEndLoc);
```

Comment:
Why do we need to check `TryEndLoc` here? The only necessary adjust is to treat cursor on`?` and `!` as if it is on `try`. Right?

Reply:
Without this placing cursor at the end of `try` in the expression like:
```swift
try throwingFunc()
```
results in an invalid cursor on current master. This check handles placing at the end of `try` keyword in any try expression (`try`, `try!`, `try?`). Two other checks are for handling placings at the end of `try!` and `try?`.

Reply:
hmm, in your added test, the behavior is to show the refactoring action in a position of `let _ = try!<CURSOR_HERE> throwingFunc()`. This is suboptimal, right? because no action should be available in white space.

Reply:
Problem is that `Lexer::getLocForStartOfToken(SM, Loc)` doesn't return the beginning of `try` as `SourceLoc` when running with this data. And since I've needed this check in two other cases (when using `try!` and `try?` and placing cursor at the end of `try`) — I've decided that it's good enough to handle the third one, too.

Reply:
Yeah, that was intentional. I think that many of the users could expect such a behaviour — cursor is still directly after the `try` and I think that conceptually it's connected to it. Whole sense of this PR is to include this placings in results, too. So if you think that it's a bad idea — we could close it.

Reply:
It's not necessarily a bad idea but an unconventional idea. Our other refactorings, such as rename, are only available when cursor lies on the relevant token... I like the other part of PR though (handling `try?`). Could you clean it?

Reply:
Yeah, you're right. After thinking more we shouldn't make one exception just for try — it could mislead the user.